c/cert/src/rules/ERR33-C/DetectAndHandleStandardLibraryErrors.ql

```diff
@@ -0,0 +1,207 @@
+          .hasName([
+              "c16rtomb", "c32rtomb", "clock", "ftell", "mbrtoc16", "mbrtoc32", "mbsrtowcs",
+              "mbstowcs", "mktime", "time", "wcrtomb", "wcsrtombs", "wcstombs"
+            ])
+      or
+      errValue.(UnaryMinusExpr).getOperand().(Literal).getValue() = "1" and
+      not this.getArgument(0) instanceof NULL and
+      this.getTarget().hasName(["mblen", "mbrlen", "mbrtowc", "mbtowc", "wctomb_s", "wctomb"])
+      or
+      errValue.getType() instanceof IntType and
+      this.getTarget().hasName(["fread", "fwrite"])
+    )
+    or
+    errOperator = ["<", ">="] and
+    (
+      errValue.(Literal).getValue() = "0" and
+      this.getTarget()
+          .hasName([
+              "fprintf_s", "fprintf", "fwprintf_s", "fwprintf", "printf_s", "snprintf_s",
+              "snprintf", "sprintf_s", "sprintf", "swprintf_s", "swprintf", "thrd_sleep",
+              "vfprintf_s", "vfprintf", "vfwprintf_s", "vfwprintf", "vprintf_s", "vsnprintf_s",
+              "vsnprintf", "vsprintf_s", "vsprintf", "vswprintf_s", "vswprintf", "vwprintf_s",
+              "wprintf_s", "printf", "vprintf", "wprintf", "vwprintf"
+            ])
+      or
+      errValue.getType() instanceof IntType and
+      this.getTarget().hasName(["strxfrm", "wcsxfrm"])
+    )
+    or
+    errOperator = "NA" and
+    (
+      errValue = any(Expr e) and
+      this.getTarget()
+          .hasName([
+              "kill_dependency", "memcpy", "wmemcpy", "memmove", "wmemmove", "strcpy", "wcscpy",
+              "strncpy", "wcsncpy", "strcat", "wcscat", "strncat", "wcsncat", "memset", "wmemset"
+            ])
+    )
+  }
+
+  Expr getErrValue() { result = errValue }
+
+  string getErrOperator() { result = errOperator }
+}
+
+// Nodes following a file write before a call to `ferror` is performed
+ControlFlowNode ferrorNotchecked(FileWriteFunctionCall write) {
+  result = write
+  or
+  exists(ControlFlowNode mid |
+    mid = ferrorNotchecked(write) and
+    //do not traverse the short-circuited CFG edge
+    not isShortCircuitedEdge(mid, result) and
+    result = mid.getASuccessor() and
+    //Stop recursion on call to ferror on the correct file
+    not accessSameTarget(result.(FerrorCall).getArgument(0), write.getFileExpr())
+  )
+}
+
+from ExpectedErrReturn err
+where
+  not isExcluded(err, Contracts5Package::detectAndHandleStandardLibraryErrorsQuery()) and
+  // calls that must be verified using the return value
+  not exists(ComparisonOperation op |
+    DataFlow::localExprFlow(err, op.getAnOperand()) and
+    (err.getErrOperator() != "NA" implies op.getOperator() = err.getErrOperator()) and
+    op.getAnOperand() = err.getErrValue() and
+    // special case for function `realloc` where the returned pointer
+    // should not be invalidated
+    not (
+      err.getTarget().hasName("realloc") and
+      op.getAnOperand().(VariableAccess).getTarget() =
+        err.getArgument(0).(VariableAccess).getTarget()
+    )
+  ) and
+  // EXCEPTIONS
```

Comment:
similarly , might be good to consider a few more predicate/classes whichever for the below bits for readability sake